### DISCUSSION TITLE
Simple master-workers example

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [sybrenjansen/mpire#100] [Q&A]
I want to understand how to transfer and execute modules on keep-alive workers. Below is a simple example:

Assume 4 cores and keep-alive workers.

Transfer the loggy.py and adder.py modules to each keep-alive worker.

From the master, call the function do_log(a=x) on each worker.

Collect the results from the workers at the master.

How do you solve this with Mpire? Thanks

```
# master.py
a = 1
d = do_log(a)
```
```
# loggy.py
import math
from adder import do_sum

def do_log(a):
    return math.log(do_sum(a))
```
```
# adder.py
def do_sum(a):
    b = 2
    return a + b
```

**Comment by sybrenjansen** [comment]
I'm not sure if I fully understand your question, but it seems you can do this with:
```
# master.py

from mpire import WorkerPool
from loggy import do_log

a = [1, 2, 3, 4]  # whatever you want, basically

with WorkerPool(n_jobs=4, keep_alive=True) as pool:
    d = pool.map(do_log, a)
    # you can reuse the workers here, if you want, because you've set keep_alive=True
    ...
```
In this example `do_log` is called by each worker once for one of the inputs in `a`. The output you get is then:
```
# output of d
 [1.0986122886681098, 1.3862943611198906, 1.6094379124341003, 1.791759469228055]
```

Does this answer your question?

**Reply by dineshbvadhia** [reply]
Thanks for the quick reply.

How do the modules loggy.py and adder.py get transferred to each worker?

**Reply by sybrenjansen** [reply]
MPIRE first starts the child processes. Then it sends the function reference (`do_log` in this case) over a queue, as would any other multiprocessing library. 

If you are using `fork` as start method (the default on unix systems) the child processes basically can make use of the same memory pool as your main process and Python can easily use your already imported function. I.e., if they are imported in the main process, the child process can use it as well. (Note that this is only true when the imports have been made before you start the child processes.)

If you're on windows or need to use 'spawn' as start method, then fresh new processes are started and imports don't get copied. Python needs to re-import them. For this to work, these functions do need to be saved in a file, which you have. Then Python can find them without issue.

**Reply by dineshbvadhia** [reply]
Assume 'spawn' for both linux and windows.

Assume there is a 'server process' on each worker to receive and respond to requests from the master.

For now, I want to know how to send loggy.py and adder.py to each worker, and to ensure the imports happen.

**Reply by sybrenjansen** [reply]
First, I'm happy to share some of the underlying workings of multiprocessing or Python in general, but I'm curious to why you're asking these questions. Is it because of curiosity, or is something not working on your end?

Anyway, for `spawn`, in this example you pass `do_log` as the function to mpire, which is actually referenced as `loggy.do_log`, as it exists in `loggy.py`. Python knows that, because you have imported it in `master.py`.

Python sends this `loggy.do_log` reference over the queue to the child processes. The child process then knows that it should import `loggy.py` and that it can find the `do_log` function there. So it imports `loggy.py`. When importing that file, it also imports `adder.py` because that line is on top of the `loggy.py`.

So it's not that you send the files to each worker. Each worker tries to find the files again on disk and imports them automatically.

**Reply by dineshbvadhia** [reply]
Thanks for the explanation. I want to determine if a working (cluster-based) ipyparallel application can be converted to Mpire to work on multiple-cores. There are many moving parts and so have to tread carefully :) 

Looking at this differently: Say, I have a bootup file to configure a worker. Can this source file be executed from the master on each worker?
